From 505eda8c731c2bdc9a904386f399899ce4fb84a8 Mon Sep 17 00:00:00 2001
From: Olivier Brunel <jjk@jjacky.com>
Date: Tue, 31 Dec 2013 13:43:26 +0100
Subject: [PATCH] treeview: Avoid calling do_validate_rows() if not needed

There's no need to always call do_validate_rows() from
get_preferred_width(), as this is only intended originally to make sure
we get a size (when not yet realized).

Additionally, this was causing issues of missing size request when
adding rows: because all rows were validated from the call in
get_preferred_width() it would turn validate_rows() call into a no-op,
except it was supposed to queue a size request.

See https://mail.gnome.org/archives/gtk-app-devel-list/2013-November/msg00049.html
for an example

With this change in, do_validate_rows() isn't called as much, and
specifically for fixed_row_height mode is isn't called from calls to
presize_handler() which in turn means that by the time do_validate_rows()
gets called, and sets the fixed_height, there's no more size request
done (so height/scroll is wrong).

We ensure the correct size is known by calling queue_resize() at the end
of initialize_fixed_height_mode() to fix this.
We can also avoid some work in do_presize_handler() if fixed_height hasn't
been set yet, since it means we don't have (correct) sizes yet.

Signed-off-by: Olivier Brunel <jjk@jjacky.com>
---
 gtk/gtktreeview.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/gtk/gtktreeview.c b/gtk/gtktreeview.c
index bf2c1b1..805af40 100644
--- a/gtk/gtktreeview.c
+++ b/gtk/gtktreeview.c
@@ -2511,8 +2511,9 @@ gtk_tree_view_get_preferred_width (GtkWidget *widget,
   /* we validate some rows initially just to make sure we have some size.
    * In practice, with a lot of static lists, this should get a good width.
    */
-  do_validate_rows (tree_view, FALSE);
-  
+  if (!gtk_widget_get_realized (widget))
+    do_validate_rows (tree_view, FALSE);
+
   tree_view->priv->minimum_width = 0;
   tree_view->priv->natural_width = 0;
   tree_view->priv->n_expand_columns = 0;
@@ -6654,6 +6655,8 @@ initialize_fixed_height_mode (GtkTreeView *tree_view)
 
    _gtk_rbtree_set_fixed_height (tree_view->priv->tree,
                                  tree_view->priv->fixed_height, TRUE);
+
+   gtk_widget_queue_resize (GTK_WIDGET (tree_view));
 }
 
 /* Our strategy for finding nodes to validate is a little convoluted.  We find
@@ -6858,7 +6861,7 @@ do_presize_handler (GtkTreeView *tree_view)
       tree_view->priv->presize_handler_tick_cb = 0;
     }
 
-  if (tree_view->priv->fixed_height_mode)
+  if (tree_view->priv->fixed_height_mode && tree_view->priv->fixed_height > 0)
     {
       GtkRequisition requisition;
 
-- 
1.9.1

